### PR TITLE
improved the efficiency of `matmul()` by completely removing concatenation

### PR DESCRIPTION
**Pull request by ParticularMiner**
- [x] Supplements #7000 
- [x] Tests passed
- [x] Passes `pre-commit run --all-files`

This PR avoids concatenation in both `blockwise()` and `reduction()` within the implementation of `matmul()`.  Previously, `matmul()` avoided concatenation only in `blockwise()`.

Due to the special nature of the preceding blockwise contraction, all chunk-operands of a sum during the reduction phase are expected to have exactly the same shape, with a size of 1 for the reduction axis.  This makes mere element-wise addition of the chunks possible.   Furthermore, the summation result can be merely reshaped to lose the reduction axis when requested.  In principle, both these operations (element-wise addition and reshaping) are performance boosters compared to the default concatenation-mediated reduction which often involves data rearrangement.

The code of `matmul()` has also been shortened, making it a bit more readable.  This is a direct consequence of ensuring that the position of the new-axis introduced in the output of chunk-function `_matmul()` \[the first argument to `blockwise()`\] is consistent with its expected position in the output of `blockwise()` (the position of the contraction-axis).

Performance is comparable to the original `matmul()` implementation in all cases except one, in which this new implementation is approximately 20% faster, namely, "the case where all dimensions are chunked".  This outcome is consistent with the total absence of concatenation-overhead in the new version.   Performance results can be found in this [jupyter notebook](https://github.com/ParticularMiner/dask/blob/matmul_performance/matmul_perf_second2last.ipynb).

See discussion with @ravwojdyla at https://github.com/dask/dask/pull/7000#discussion_r756926031 for the motivation of this PR.

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by ParticularMiner** [comment]
Attention:  @dask, admins

Many CI tests are broken (but not because of this PR), as seen here (#8423) and elsewhere (#8424).

**Comment by ParticularMiner** [comment]
Hi @dask

I need some help debugging this PR for the unit-test `dask/array/tests/test_array_function.py::test_unregistered_func[<lambda>6] `, which passes with numpy version 1.21.4, but not versions &le; 1.19.5.  

The bug originates from using a list of arrays of type `dask.array.tests.test_dispatch.EncapsulateNDArray` in`numpy.sum` which is implemented by `numpy.add.reduce()`.  I have not been able to determine what changed in `numpy.add.reduce()` between versions 1.19.5 and 1.21.4 that killed this bug.  The error thrown is 

`ValueError: setting an array element with a sequence.`

Traceback log is as follows:

```python
=================================== FAILURES ===================================
______________________ test_unregistered_func[<lambda>6] _______________________
TypeError: float() argument must be a string or a number, not 'EncapsulateNDArray'

The above exception was the direct cause of the following exception:

func = <function <lambda> at 0x7f0330d56560>

    @pytest.mark.parametrize(
        "func",
        [
            lambda x: np.concatenate([x, x, x]),
            lambda x: np.cov(x, x),
            lambda x: np.dot(x, x),
            lambda x: np.dstack((x, x)),
            lambda x: np.flip(x, axis=0),
            lambda x: np.hstack((x, x)),
            lambda x: np.matmul(x, x),
            lambda x: np.mean(x),
            lambda x: np.stack([x, x]),
            lambda x: np.sum(x),
            lambda x: np.var(x),
            lambda x: np.vstack((x, x)),
            lambda x: np.linalg.norm(x),
        ],
    )
    def test_unregistered_func(func):
        # Wrap a procol-based encapsulated ndarray
        x = EncapsulateNDArray(np.random.random((100, 100)))
    
        # See if Dask holds the array fine
        y = da.from_array(x, chunks=(50, 50))
    
        # Check if it's an equivalent array
        assert_eq(x, y, check_meta=False, check_type=False)
    
        # Perform two NumPy functions, one on the
        # Encapsulated array
        xx = func(x)
    
        # And one on the Dask array holding these
        # encapsulated arrays
        yy = func(y)
    
        # Check that they are equivalent arrays.
>       assert_eq(xx, yy, check_meta=False, check_type=False)

dask/array/tests/test_array_function.py:177: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/array/utils.py:283: in assert_eq
    b, check_shape=check_shape, check_graph=check_graph, check_chunks=check_chunks
dask/array/utils.py:242: in _get_dt_meta_computed
    x = _check_chunks(x)
dask/array/utils.py:218: in _check_chunks
    x = x.persist(scheduler="sync")
dask/base.py:261: in persist
    (result,) = persist(self, traverse=False, **kwargs)
dask/base.py:833: in persist
    results = schedule(dsk, keys, **kwargs)
dask/local.py:558: in get_sync
    **kwargs,
dask/local.py:496: in get_async
    for key, res_info, failed in queue_get(queue).result():
/usr/share/miniconda3/envs/test-environment/lib/python3.7/concurrent/futures/_base.py:428: in result
    return self.__get_result()
/usr/share/miniconda3/envs/test-environment/lib/python3.7/concurrent/futures/_base.py:384: in __get_result
    raise self._exception
dask/local.py:538: in submit
    fut.set_result(fn(*args, **kwargs))
dask/local.py:234: in batch_execute_tasks
    return [execute_task(*a) for a in it]
dask/local.py:234: in <listcomp>
    return [execute_task(*a) for a in it]
dask/local.py:225: in execute_task
    result = pack_exception(e, dumps)
dask/local.py:220: in execute_task
    result = _execute_task(task, data)
dask/core.py:119: in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
dask/array/routines.py:344: in _chunk_sum
    out = chunk.sum(a, dtype=dtype, axis=0)
<__array_function__ internals>:6: in sum
    ???
/usr/share/miniconda3/envs/test-environment/lib/python3.7/site-packages/numpy/core/fromnumeric.py:2229: in sum
    initial=initial, where=where)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = [<dask.array.tests.test_dispatch.EncapsulateNDArray object at 0x7f0331e7c710>, <dask.array.tests.test_dispatch.EncapsulateNDArray object at 0x7f0330508dd0>]
ufunc = <ufunc 'add'>, method = 'sum', axis = 0, dtype = dtype('float64')
out = None
kwargs = {'initial': <no value>, 'keepdims': <no value>, 'where': <no value>}
passkwargs = {}

    def _wrapreduction(obj, ufunc, method, axis, dtype, out, **kwargs):
        passkwargs = {k: v for k, v in kwargs.items()
                      if v is not np._NoValue}
    
        if type(obj) is not mu.ndarray:
            try:
                reduction = getattr(obj, method)
            except AttributeError:
                pass
            else:
                # This branch is needed for reductions like any which don't
                # support a dtype.
                if dtype is not None:
                    return reduction(axis=axis, dtype=dtype, out=out, **passkwargs)
                else:
                    return reduction(axis=axis, out=out, **passkwargs)
    
>       return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
E       ValueError: setting an array element with a sequence.

/usr/share/miniconda3/envs/test-environment/lib/python3.7/site-packages/numpy/core/fromnumeric.py:90: ValueError
```


Many thanks in advance for your help!

**Comment by ParticularMiner** [comment]
Yay! CI tests are now operational.  Thanks @dask admin!

**Comment by jsignell** [comment]
Sorry for the lack of response @ParticularMiner. I'll leave review to @ravwojdyla or maybe @eriknw

File: dask/array/routines.py [hunk 1]
```diff
@@ -332,6 +333,52 @@ def vdot(a, b):
...
+    # the output can be merely reshaped to lose the `axis`-
+    # dimension when keepdims = False
+    if type(a) is list:
+        item_type = type(a[0])
+        if item_type is np.ndarray:
+            out = chunk.sum(a, dtype=dtype, axis=0)
+        else:
+            a = [item.__array__() for item in a]
+            out = chunk.sum(a, dtype=dtype, axis=0)
+            out = item_type(out)
```
Comment (ravwojdyla) [inline dask/array/routines.py:358]: I'm probably missing something, but why does `chunk.sum` doesn't work here? could we let numpy delegate to the array types? `__array__` might lead to numpy array materialisation right, which can be expensive? afair CuPy doesn't support `__array__`.

**Review by ravwojdyla — CHANGES_REQUESTED** [review]
great work @ParticularMiner, see some questions/comments below.

File: dask/array/routines.py [hunk 2]
```diff
@@ -332,6 +333,52 @@ def vdot(a, b):
...
+    else:
+        return out.reshape(_shape_minus_axis(out.shape, axis[0]))
+
+
+def _sum_wo_cat(a, axis=None, dtype=None):
+    if dtype is None:
+        dtype = getattr(np.zeros(1, dtype=a.dtype).sum(), "dtype", object)
+
+    if a.shape[axis] == 1:
+        return a.reshape(_shape_minus_axis(a.shape, axis))
```
[the remark this answers is not in the document]
Comment (ParticularMiner) [reply dask/array/routines.py:374]: Aha!  You're precisely right.  I forgot all about `squeeze`.  I will change that.

File: dask/array/routines.py [hunk 3]
```diff
@@ -332,6 +333,52 @@ def vdot(a, b):
...
+
+    result = reduction(
```
[the remark this answers is not in the document]
Comment (ParticularMiner) [reply dask/array/routines.py:376]: Of course.  Sorry for that.

File: dask/array/routines.py [hunk 4]
```diff
@@ -332,6 +333,52 @@ def vdot(a, b):
     return dot(a.conj().ravel(), b.ravel())
 
 
+def _shape_minus_axis(shape, axis):
```
[the remark this answers is not in the document]
Comment (ParticularMiner) [reply dask/array/routines.py:336]: Indeed. I'll be using `squeeze` instead.  Thanks for this.

File: dask/array/routines.py [hunk 5]
```diff
@@ -332,6 +333,52 @@ def vdot(a, b):
...
+    newshape = list(shape)
+    newshape.pop(axis)
+    return tuple(newshape)
+
+
+def _chunk_sum(a, axis=None, dtype=None, keepdims=None):
```
[the remark this answers is not in the document]
Comment (ParticularMiner) [reply dask/array/routines.py:342]: I'll have to check the debugger to see if `dtype` is actually being passed or not.  I'll let you know.

File: dask/array/routines.py (hunk 1 above)
Comment (ParticularMiner) [reply dask/array/routines.py:358]: I agree that there's a problem here.  But I'm not sure how to fix it properly.  

It turns out that `chunk.sum`  works for the latest version of numpy so then this conditional branch is unnecessary.  But the CI tests revealed that earlier versions of numpy raise a `ValueError: setting an array element with a sequence.`

I'll explore further.  Thanks.

File: dask/array/routines.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ParticularMiner) [reply dask/array/routines.py:342]: Ok.  So my debugger says `_execute_task()` defined in core.py is calling this function and is passing `dtype`.  

I believe this was setup within the task graph by `reduction` which in turn was called by `sum_wo_cat`.  See the return statement in `sum_wo_cat`:

```python
    return reduction(
        a, _chunk_sum, _chunk_sum, axis=axis, dtype=dtype, concatenate=False
    )
```

File: dask/array/routines.py (hunk 1 above)
Comment (ParticularMiner) [reply dask/array/routines.py:358]: I forgot to mention &mdash; the specific chunk type that raises the above exception is `dask.array.tests.test_dispatch.EncapsulateNDArray` found defined in dask/array/tests/test_dispatch.py

**Comment by ParticularMiner** [comment]
@ravwojdyla 

Using `squeeze()` would also be my preference.  

Unfortunately, again, for the array type `dask.array.tests.test_dispatch.EncapsulateNDArray` (found defined in [dask/array/tests/test_dispatch.py](https://github.com/dask/dask/blob/a5aecac8313fea30c5503f534c71f325b1775b9c/dask/array/tests/test_dispatch.py#L36)) `squeeze()`  raises an `AttributeError: 'EncapsulateNDArray' object has no attribute 'squeeze'`.

So I reluctantly have to revert to using `reshape()` instead.  Pity!

**Comment by ParticularMiner** [comment]
@ravwojdyla 

I have now replaced the following code in `_chunk_sum()`:
```python
        item_type = type(a[0])
        if item_type is np.ndarray:
            out = chunk.sum(a, dtype=dtype, axis=0)
        else:
            a = [item.__array__() for item in a]
            out = chunk.sum(a, dtype=dtype, axis=0)
            out = item_type(out)
```
with this
```python
        xp = np

        if is_cupy_type(a[0]):
            import cupy

            xp = cupy

        out = reduce(partial(xp.add, dtype=dtype), a)
```

where reduce is imported from the package `functools`.   Fortunately, all CI tests pass with these changes.  Moreover, just like in `_matmul()` I explicitly test for the existence of a `cupy` type.

The latter code snippet seems to give a slight performance edge over the former making it more than 25% faster (compared to 20% by the former) than than the original `matmul` in the case where all dimensions are chunked.

File: dask/array/routines.py (hunk 1 above)
Comment (ravwojdyla) [reply dask/array/routines.py]: I think you need to add `squeeze` to `EncapsulateNDArray`, see https://github.com/dask/dask/blob/a5aecac8313fea30c5503f534c71f325b1775b9c/dask/array/tests/test_dispatch.py#L84-L87

If we fix that what else is missing?

File: dask/array/routines.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ravwojdyla) [reply dask/array/routines.py:342]: Ah, interesting, thanks! The doc probably needs an update then 🤔 : https://github.com/dask/dask/blob/a5aecac8313fea30c5503f534c71f325b1775b9c/dask/array/reductions.py#L122-L138

Obviously out of context for this PR.

File: dask/array/routines.py [hunk 6]
```diff
@@ -332,6 +333,52 @@ def vdot(a, b):
...
+
+        if is_cupy_type(a[0]):
+            import cupy
+
+            xp = cupy
+
+        out = reduce(partial(xp.add, dtype=dtype), a)
+    else:
+        if a.shape == (0,):
+            return a
```
Comment (ravwojdyla) [inline dask/array/routines.py:362]: What is this case? And does it need to be a separate branch?

File: dask/array/routines.py (hunk 1 above)
Comment (ParticularMiner) [reply dask/array/routines.py]: I wanted to do that earlier but hesitated because got the impression the creators wanted the `EncapsulateNDArray` interface to remain minimal as mentioned in the docstring for `WrappedArray` in [dask/dask/array/tests/test_dispatch.py](https://github.com/dask/dask/blob/a5aecac8313fea30c5503f534c71f325b1775b9c/dask/array/tests/test_dispatch.py#L93):
```python
class WrappedArray(np.lib.mixins.NDArrayOperatorsMixin):
    """
    Another mock duck array class (like EncapsulateNDArray), but
    designed to be above Dask in the type casting hierarchy (that is,
    WrappedArray wraps Dask Array) and be even more minimal in API.
    Tests that Dask defers properly to upcast types.
    """
``` 

Perhaps you feel that's not critical?

File: dask/array/routines.py [hunk 7]
```diff
@@ -332,6 +333,52 @@ def vdot(a, b):
...
+    # ment-wise addition of the arrays possible.   Besides,
+    # the output can be merely reshaped to lose the `axis`-
+    # dimension when keepdims = False
+    if type(a) is list:
+        xp = np
+
+        if is_cupy_type(a[0]):
+            import cupy
+
+            xp = cupy
```
Comment (ravwojdyla) [inline dask/array/routines.py:357]: Why is this needed? wouldn't numpy handle the dispatch to cupy below?

File: dask/array/routines.py [hunk 8]
```diff
@@ -332,6 +333,52 @@ def vdot(a, b):
...
+
+        out = reduce(partial(xp.add, dtype=dtype), a)
+    else:
+        if a.shape == (0,):
+            return a
+        out = a
+    if keepdims:
+        return out
+    else:
+        return out.reshape(_shape_minus_axis(out.shape, axis[0]))
```
Comment (ravwojdyla) [inline dask/array/routines.py:367]: We can probably make the squeeze work (see the other comment), but otherwise if this method is used only once, probably not worth a complete new method.

**Review by ravwojdyla — CHANGES_REQUESTED** [review]

File: dask/array/routines.py (hunk 6 above)
Comment (ParticularMiner) [reply dask/array/routines.py:362]: This branch turns out to be necessary, as `_chunk_sum` is often called with an empty array `a` parameter passed in order to compute meta data &mdash; a peculiarity of `dask.array.reduction`

File: dask/array/routines.py (hunk 1 above)
Comment (ravwojdyla) [reply dask/array/routines.py]: It is my **believe**, that if we go forward with using `squeeze` in `matmul` (which arguable is a pretty common op), adding `squeeze` to that test `WrappedArray` is "justified".

File: dask/array/routines.py (hunk 7 above)
Comment (ParticularMiner) [reply dask/array/routines.py:357]: Honestly, I do not know.  But the exact same construct is being used in `_matmul()` (which you allowed to remain), which is why I copied it over here.

File: dask/array/routines.py (hunk 6 above)
Comment (ravwojdyla) [reply dask/array/routines.py:362]: @ParticularMiner if you remove that branch and let it go into the other branch, eventually into the `reduce`/`add`, what happens, how does it fail?

File: dask/array/routines.py (hunk 8 above)
Comment (ParticularMiner) [reply dask/array/routines.py:367]: I agree.  Let me know, after you read my response to adding a squeeze method to `EncapsulatedNDArray`, what you think.

File: dask/array/routines.py (hunk 7 above)
Comment (ravwojdyla) [reply dask/array/routines.py:357]: @ParticularMiner that part actually came after my change via https://github.com/dask/dask/pull/7563. @quasiben would appreciate your input here? When is that `is_cupy_type` needed?

File: dask/array/routines.py (hunk 6 above)
Comment (ParticularMiner) [reply dask/array/routines.py:362]: I get an `IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed` 
at the `reduce`/`add` line when computing the meta data.   Would you like the full traceback log?

Comment (ParticularMiner) [reply dask/array/routines.py:362]: Here's the error log from one test:
[tracebacklog.txt](https://github.com/dask/dask/files/7645059/tracebacklog.txt)

Comment (ParticularMiner) [reply dask/array/routines.py:362]: The problem is that the error occurs in dask *outside* any of the code that I wrote in this PR.

File: dask/array/routines.py (hunk 7 above)
Comment (ParticularMiner) [reply dask/array/routines.py:357]: Sorry I assumed it was yours @ravwojdyla

File: dask/array/routines.py (hunk 6 above)
Comment (ravwojdyla) [reply dask/array/routines.py:362]: @ParticularMiner thanks, oh, so this error happens only for a specific case: `[(7,), (7,), (), ()],`, which will reduce to a  scalar, are you sure there isn't something going on here? 🤔

File: dask/array/routines.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ParticularMiner) [reply dask/array/routines.py:342]: Seems you're right &mdash; the docs need to be updated.  Sadly, I find the dask docs to be somewhat scanty.  Most of what I've learned about dask was through trial and error.  ☹️ 

Still I appreciate the fact that writing docs for a program like dask could be daunting.

File: dask/array/routines.py (hunk 6 above)
Comment (ParticularMiner) [reply dask/array/routines.py:362]: Oh no.  It's not the only case.  I just uploaded the error for one case. There are at least 15 other cases where the error occurs for that test unit `test_matmul()`.

File: dask/array/routines.py (hunk 1 above)
Comment (ParticularMiner) [reply dask/array/routines.py]: True.  So should I add it and revert back to using `numpy.sum` instead of `numpy.add` with `functools.reduce`?  Is there anyone who'd be unhappy with that?

**Comment by ParticularMiner** [comment]
Thanks for your detailed review on this @ravwojdyla !  I really appreciate that.

It's getting to midnight here, so I'll likely respond to any more messages tomorrow.

File: dask/array/routines.py (hunk 1 above)
Comment (ravwojdyla) [reply dask/array/routines.py]: I'm +1, I would be interested to see the impact there, also to the other comment: https://github.com/dask/dask/pull/8423#discussion_r761478288

**Comment by ParticularMiner** [comment]
@ravwojdyla 

You will find that I added the `squeeze()` method to `EncapsulateNDArray`, discarded `shape_minus_axis` together with one conditional branch of `_chunk_sum`.  Everything seems to work well with these changes.  Performance seems to be even somewhat faster than before. 

I guess the only outstanding issue is the one regarding the `cupy` type array.

Thanks.

**Comment by ravwojdyla** [comment]
@ParticularMiner perfect, thank you. I **suspect** that `cupy` branch is not needed, but there's a chance I don't fully understand why it would be there, maybe let's give @quasiben some time to respond to the comment in https://github.com/dask/dask/pull/8423#discussion_r761452511

**Comment by ParticularMiner** [comment]
Sure @ravwojdyla. Thanks for the great suggestions.

> I **suspect** that `cupy` branch is not needed, but there's a chance I don't fully understand why it would be there, maybe let's give @quasiben some time to respond to the comment in [#8423 (comment)](https://github.com/dask/dask/pull/8423#discussion_r761452511)

It seems @jakirkham might also be able to address this question:

As `dask.array` is right now, is it necessary to test for the `cupy` chunk type (as @quasiben introduced:
https://github.com/dask/dask/blob/43247806aa385d6f958613f77a5b84c3a763823e/dask/array/routines.py#L335-L347 for `dask.array.matmul` in #7563) whenever one needs to use the `cupy` implementation of some `numpy` function like `numpy.add` or `numpy.sum`?
Is there a reason it would be less convenient to use a more centralized approach, like register the particular `cupy` function implementation in a "lookup registry", as has been done, for example, with `concatenate` for various chunk types including `cupy` itself?

https://github.com/dask/dask/blob/43247806aa385d6f958613f77a5b84c3a763823e/dask/array/backends.py#L113-L129

**Review by ravwojdyla — APPROVED** [review]
Approving*, great work @ParticularMiner

\* pending resolution of the question about the need for `cupy` check ( https://github.com/dask/dask/pull/8423#discussion_r761439317). @ParticularMiner that said if you prefer to merge sooner, and open an issue to discuss that branch/question etc, that sounds good to me too.

FYI @jsignell (and thanks for the ping).

**Comment by ParticularMiner** [comment]
That's fine, @ravwojdyla .  It's not urgent to merge now.

> * pending resolution of the question about the need for `cupy` check ( [#8423 (comment)](https://github.com/dask/dask/pull/8423#discussion_r761439317)). @ParticularMiner that said if you prefer to merge sooner, and open an issue to discuss that branch/question etc, that sounds good to me too.

**Comment by jsignell** [comment]
@ravwojdyla thank you so much for reviewing this work and thank you @ParticularMiner for engaging! I'll broaden the ping on https://github.com/dask/dask/pull/8423#discussion_r761439317 a bit to see if anyone on @dask/gpu knows if that branch is still necessary.

File: dask/array/routines.py (hunk 7 above)
Comment (pentschev) [reply dask/array/routines.py:351]: `is_cupy_type` is normally needed when we need to support the non-NumPy API part of CuPy, such as `cupyx.scipy`, this is why `matmul` requires that check. I think the reduction here may not need the same, or at least I can't think of a need for that right now. I would then suggest just replacing all this by `np.add` and it should work. If gpuCI tests pass we should be good to what we know is supported today.

@quasiben please feel free to correct me.

File: dask/array/routines.py [hunk 9]
```diff
@@ -332,6 +333,45 @@ def vdot(a, b):
...
+    # dimension when keepdims = False
+    if type(a) is list:
+        xp = np
+
+        if is_cupy_type(a[0]):
+            import cupy
+
+            xp = cupy
+
+        out = reduce(partial(xp.add, dtype=dtype), a)
```
Comment (pentschev) [inline dask/array/routines.py:353]:
```suggestion
        out = reduce(partial(np.add, dtype=dtype), a)
```

Comment (ParticularMiner) [reply dask/array/routines.py:353]: Thank you @pentschev 

One final question: what about the following code-block already in the dask source-code (but also directly related to this PR)?
```python
def _matmul(a, b):
    xp = np

    if is_cupy_type(a):
        import cupy
        xp = cupy

    chunk = xp.matmul(a, b)
```
Can it also be similarly replaced as you did?
```python
    chunk = np.matmul(a, b)
```

Comment (ravwojdyla) [reply dask/array/routines.py]: @ParticularMiner I **believe** it could be removed **BUT** only assuming certain cupy version (>=9.0), see https://github.com/cupy/cupy/issues/4549. I assume that assumption is not "safe", 9.0 was relatively recent. I think if I was you, would add a comment there to the cupy issue + this discussion + explanation that it probably can be eventually removed(?) Unless of course we dask is comfortable with failing there and forcing cupy upgrade, which is probably not worth the potential bad user experience.

Comment (ParticularMiner) [reply dask/array/routines.py]: @ravwojdyla Very interesting link!  I wasn't aware of the `__array_ufunc__` protocol.  Yes, I completely agree with you that we should beware of possible backward compatibility issues here.  So I'll leave the code as it is and include a comment, as you suggest.

**Comment by ParticularMiner** [comment]
@ravwojdyla 
Thanks again for clearing things up!

@jsignell 
It appears all outstanding issues here have been addressed, and this PR is now awaiting your deployment.
Many thanks for your effective oversight!

**Comment by ParticularMiner** [comment]
I believe the observed test-failures for Python 3.7 on Windows are not related to this PR.  (Correct me if I'm wrong.)

**Comment by jsignell** [comment]
I agree that the failing tests are unrelated. They are also showing up on main.

**Comment by jsignell** [comment]
Thanks for this work! The end diff looks super clean :)

File: dask/array/routines.py (hunk 7 above)
Comment (ParticularMiner) [reply dask/array/routines.py]: Sorry @pentschev , I didn't see your comment above until just now.  Nevertheless, I've already taken steps that agree with your comment.
